### PR TITLE
[SPARK-31582] [Yarn] Being able to not populate Hadoop classpath

### PR DESCRIPTION
**Pull request by dbtsai**
### What changes were proposed in this pull request?
We are adding a new Spark Yarn configuration, `spark.yarn.populateHadoopClasspath` to not populate Hadoop classpath from `yarn.application.classpath` and `mapreduce.application.classpath`.

### Why are the changes needed?
Spark Yarn client populates extra Hadoop classpath from `yarn.application.classpath` and `mapreduce.application.classpath` when a job is submitted to a Yarn Hadoop cluster.

However, for `with-hadoop` Spark build that embeds Hadoop runtime, it can cause jar conflicts because Spark distribution can contain different version of Hadoop jars.

One case we have is when a user uses an Apache Spark distribution with its-own embedded hadoop, and submits a job to Cloudera or Hortonworks Yarn clusters, because of two different incompatible Hadoop jars in the classpath, it runs into errors.

By not populating the Hadoop classpath from the clusters can address this issue.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
An UT is added, but very hard to add a new integration test since this requires using different incompatible versions of Hadoop.

We also manually tested this PR, and we are able to submit a Spark job using Spark distribution built with Apache Hadoop 2.10 to CDH 5.6 without populating CDH classpath.

### CODE REVIEW

**Comment by dbtsai** [comment]
cc @vanzin @srowen @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @dbtsai .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Since the default is still `true`, this is a safe configuration for the described use case.

**Comment by srowen** [comment]
I generally don't like yet another config option, but I don't object to this.
Is there any simpler way to solve it?

**Comment by dbtsai** [comment]
The other way we can do is changing `yarn.application.classpath` and `mapreduce.application.classpath` to an empty folder, but this is more hackey unfortunately.

**Comment by dbtsai** [comment]
Ping @jerryshao @tgravescs @ulysses-you who work on this area for more eyes.

**Comment by viirya** [comment]
One question, for the case mentioned in the description, "One case we have is when a user uses an Apache Spark distribution with its-own embedded hadoop, and submits a job to Cloudera or Hortonworks Yarn clusters", since the embedded hadoop is incompatible to the cluster, is it generally okay to submit the app there? Even you don't populate the classpath, will RPC or protocol be a problem?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala
Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala:74]: we might add a comment saying requires bundling Hadoop with spark or user to provide Hadoop install separately.
Also need to add this to configuration.md doc

**Comment by tgravescs** [comment]
overall  I think the change is fine. You could end up having issues if the user tried the external shuffle service and things weren't compatible but that is problem with anyone trying to run multiple versions.  obviously requires them to have all the dependencies they need.

@viirya  its ok to have multiple versions of spark on a cluster and many times there are just dependency issues with the different Hadoop versions. It will be rpc compatible but you run into different versions of dependencies (guava, jetty, etc). 

The other way to possible solve this is to make sure these versions are on the class path first, but that can get tricky.

Is the standard spark package that includes Hadoop sufficient to run for you then @dbtsai ?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala
Comment (dongjoon-hyun) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala:74]: +1 for the comment and the doc.

**Comment by viirya** [comment]
@tgravescs I see. That makes sense. So users still need to pay attention on dependencies issue if any when running multiple versions there.

**Comment by dbtsai** [comment]
@tgravescs the standard "with hadoop" Spark built with Yarn is sufficient to work in our CDH env. Thanks for the review. I'll update the doc and config.md.

**Comment by adamkennedy** [comment]
@viirya The embedded Hadoop could easily be incompatible with the cluster. But then the Spark version could be incompatible with the cluster's Hadoop as well. If the Spark chooses to bundle Hadoop libraries, then at least we know that works, and then the onus is on the Spark distribution to provide a different-but-compatible version, and the failure is only a failure for that one job. But at least the Spark provider gets an opportunity to provide something working, as opposed to the situation where different parts of the two different distributions end up cross-loading each other unexpectedly.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
@dbtsai @adamkennedy Thanks for clarifying that. Making sense.

**Review by tgravescs — APPROVED** [review]
changes look fine, assuming you are pulling back into 2.4.6, otherwise version needs to be updated.

**Comment by dongjoon-hyun** [comment]
cc @holdenk since this seems to be targeting 2.4.6.

**Review by holdenk — COMMENTED** [review]
Thanks for looping me in on this I’ll hold off on cutting an 2.4.6 RC for now

**Comment by dbtsai** [comment]
@tgravescs I plan to bring this to 2.4.6 since it's default to true and there is no risk.

Many Spark users who use vendor's hadoop such as CDH or HDP are stuck on very old versions of Spark. When they try to upgrade the Spark version, vendors ask them to upgrade the whole stack which is often not realistic. With this feature, those users will be able to migrate to latest OSS Apache Spark without depending on vendors.

**Comment by dbtsai** [comment]
Thanks all for reviews. I added a new UT, and after test passes, I'll merge it into master, branch-3.0, branch-2.4.

**Comment by dbtsai** [comment]
https://github.com/apache/spark/pull/28788 is a PR that truns `spark.yarn.populateHadoopClasspath` to `false` by default for `with-hadoop` build.

**Pull request closed without merging**
